firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/LiveSession.kt

```diff
@@ -0,0 +1,359 @@
+    internal class Internal(val realtimeInput: MediaChunks) {
+      @Serializable internal data class MediaChunks(val mediaChunks: List<MediaData.Internal>)
+    }
+    fun toInternal() = Internal(Internal.MediaChunks(mediaChunks))
+  }
+
+  internal data class ToolCallSetup(
+    val functionCalls: List<FunctionCallPart.Internal.FunctionCall>
+  ) {
+
+    @Serializable
+    internal class Internal(val toolCall: ToolCall) {
+
+      @Serializable
+      internal data class ToolCall(val functionCalls: List<FunctionCallPart.Internal.FunctionCall>)
+    }
+
+    fun toInternal(): Internal {
+      return Internal(Internal.ToolCall(functionCalls))
+    }
+  }
+
+  /**
+   * Receives all function call responses from the server for the audio conversation feature.
+   *
+   * @return A [Flow] which will emit list of [FunctionCallPart] as they are returned by the model.
+   */
+  public fun receiveAudioConversationFunctionCalls(): Flow<List<FunctionCallPart>> {
+    return functionCallChannel.receiveAsFlow()
+  }
+
+  private fun fillRecordedAudioQueue() {
+    CoroutineScope(backgroundDispatcher).launch {
+      audioHelper!!.startRecording().collect {
+        if (!isRecording) {
+          cancel()
+        }
+        audioQueue.add(it)
+      }
+    }
+  }
+
+  private suspend fun sendAudioDataToServer() {
+    var offset = 0
+    val audioBuffer = ByteArray(MIN_BUFFER_SIZE * 2)
+    while (isRecording) {
+      val receivedAudio = audioQueue.poll() ?: continue
+      receivedAudio.copyInto(audioBuffer, offset)
+      offset += receivedAudio.size
+      if (offset >= MIN_BUFFER_SIZE) {
+        sendMediaStream(listOf(MediaData("audio/pcm", audioBuffer)))
+        audioBuffer.fill(0)
+        offset = 0
+      }
+    }
+  }
+
+  private fun fillServerResponseAudioQueue() {
+    CoroutineScope(backgroundDispatcher).launch {
+      receive(listOf(ContentModality.AUDIO)).collect {
+        if (!isRecording) {
+          cancel()
+        }
+        when (it.status) {
+          LiveContentResponse.Status.INTERRUPTED ->
+            while (!playBackQueue.isEmpty()) playBackQueue.poll()
+          LiveContentResponse.Status.NORMAL ->
+            if (!it.functionCalls.isNullOrEmpty()) {
+              functionCallChannel.send(it.functionCalls)
+            } else {
+              val audioData = it.data?.parts?.get(0)?.asInlineDataPartOrNull()?.inlineData
+              if (audioData != null) {
+                playBackQueue.add(audioData)
+              }
+            }
+        }
+      }
+    }
+  }
+
+  private fun playServerResponseAudio() {
+    CoroutineScope(backgroundDispatcher).launch {
+      while (isRecording) {
+        val x = playBackQueue.poll() ?: continue
+        audioHelper?.playAudio(x)
+      }
+    }
+  }
+
+  /**
+   * Starts an audio conversation with the Gemini server, which can only be stopped using
+   * stopAudioConversation.
+   */
+  public suspend fun startAudioConversation() {
+    if (isRecording) {
+      return
```

Comment:
Should we also log a warning here that it was called again?

Reply:
done